File: 49-group-anagrams/49-group-anagrams.java
import java.util.*;
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        //approach one 
        Map<String,List<String>> hm = new HashMap<>();
        List<List<String>> res = new ArrayList<>();
        for(int i = 0; i<strs.length; i++) {
            char[] temp = strs[i].toCharArray();
            Arrays.sort(temp);
            String k = new String(temp);
            if(hm.containsKey(k))
            {
                hm.get(k).add(strs[i]);
            }
            else
                hm.put(k,new ArrayList(Arrays.asList(strs[i])));
        }
        for(List<String> val:hm.values()) {
            res.add(val);
        }
        return res;
    }
}

File: 49-group-anagrams/49-group-anagrams.py
# Approach used: using count of char as key and its value as list of strings 
# Time complexity: O(MN)
# Space complexity: O(MN)

from collections import defaultdict
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        res = defaultdict(list)
        for s in strs:
            count = [0] * 26 

            for c in s:
                count[ord(c) - ord("a")] +=1 
            res[tuple(count)].append(s)
        return res.values()
